content/blog/2020-12-18-type-and-effect-system.md

```diff
@@ -0,0 +1,93 @@
++++
+title = "A Semi-Practical Type-and-Effect System"
+[extra]
+bio = """
+  Kenneth Li is a first-semester M.Eng. student who is interested in functional
+  programming. He sings bass in the 
+  [Cornell Hangovers](http://www.hangovers.com/).
+"""
+[[extra.authors]]
+name = "Kenneth Li"
++++
+
+# Introduction
+Side effects in programming are any observable effects caused by executing some code beyond its return values. For example, the simple `print` functions inherent in most languages cause the side effect of printing to the standard output stream. However, because side effects can be observed from the outside, such as by the program user, then to enforce determinism of programs the order of side-effect-causing instructions must be the same. However, side-effect-free instructions, which we will refer to as "pure", can be rearranged arbitrarily or even precomputed, so long as the code results in the same output. Unfortunately, in most modern programming languages, it is difficult to know what side effects are caused by what code, making it difficult for compilers to reorder instructions. In this report, we explore a way to encode these side effects in a static manner.
+
+In order to perform side-effect-aware static analysis, we introduce the idea of a _type-and-effect system_, which extends the notion of types to encode side effect information. For example, the OCaml function `print_int` has type `int -> unit`, but in a type-and-effect system, the signature might look more like `print_int : (int -> unit !output)`. The following section explores one implementation of such a system, available [here](https://github.com/kq-li/type-and-effect-system).
```

Comment:
FWIW, "we introduce" sounds like you invented the idea. (I know you didn't mean to, but people often say stuff like "we introduce a foo" to mean "this paper is about a new foo we made" in papers.) Linking to/citing some existing work on type-and-effect systems might be cool too!

Reply:
makes sense!